[ClangImporter] Clean up adjustTypeForConcreteImport

jrose-apple opened the pull request:
A handful of cleanups that make ImportType's `adjustTypeForConcreteImport` easier to follow. No intended functionality change.

jrose-apple committed:
[ClangImporter] Clean up adjustTypeForConcreteImport, part 1

Use a switch over the import hint to cover most of the special cases.
No functionality change.

jrose-apple committed:
[ClangImporter] Add ImportHint::VAList instead of rederiving that info

We already know something's a va_list when we import it; no need to
check again later.

jrose-apple committed:
[ClangImporter] Clean up adjustTypeForConcreteImport, part 2

(or part 4, depending on how you want to count)

Move the out-parameter logic into the switch over import hints (under
the OtherPointer case), and move the helper lambdas out into their own
functions. They're not called from anywhere else, but
adjustTypeForConcreteImport is long as it is.

Additionally, avoid early returns in the import hint switch; while
early returns are generally a good thing, having some cases
early-return and not others seems inconsistent. The post-switch logic
is pretty simple and easy to early-exit now.

No intended functionality change.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
\*grumble\* Rebased incorrectly.

jrose-apple committed:
[ClangImporter] Make it easier to get Swift pointer types

No functionality change. Nice to drop a string-based API while I'm at
it.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -688,6 +693,10 @@ namespace {
           break;
         }
 
+        static llvm::StringLiteral vaListNames[] = {
```
Oops, I meant to make this `const`. It'll probably be optimized as such anyway, but still.

jrose-apple committed:
[ClangImporter] Eliminate ImportHint::SwiftNewtypeFromCFPointer

It's no longer substantially different from ImportHint::CFPointer, so
just handle the two cases together. No functionality change.

jrose-apple commented:
Okay, so the Linux failures are due to the Linux stdlib not having AutoreleasingUnsafeMutablePointer. Previously the importer would just silently drop these decls. @compnerd, what do you want to do about this?

1. Continue silently dropping types that involve AutoreleasingUnsafeMutablePointer on non-Apple platforms.
2. Disable tests that exercise AutoreleasingUnsafeMutablePointer on Linux.
3. Look for AutoreleasingUnsafeMutablePointer in the ObjectiveC module if it can't be found in the stdlib, for testing purposes on non-Apple platforms.
4. Something else?

I'm leaning towards #2; the only reason we enabled ObjC tests on non-Apple platforms was to make it easier for non-Apple people to notice if they accidentally broke something, right?

jrose-apple commented:
Hm, although I guess some of these are failing almost coincidentally, since the `id __unsafe_unretained *` isn't actually being used. (Hence how we got away with it before.) I guess I'll leave it as is for now.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Uh, hm, that got some but not all better.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@rintaro, do you know what's up with this Syntax failure? I don't see how my ClangImporter changes could be affecting Syntax.

rintaro commented:
I'll look into it.

brentdax commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -60,6 +60,36 @@ bool ClangImporter::Implementation::isOverAligned(clang::QualType type) {
+  if (!pointerDecl)
+    return Type();
+
+  return BoundGenericType::get(pointerDecl, /*parent*/nullptr, pointee);
```
It looks like lib/Sema/TypeCheckType.cpp has a `static Type getPointerType()` helper that looks a lot like this one. Think it's worth extracting a method to a common location? Maybe `Type Type::getPointerType(PointerTypeKind kind)`?

jrose-apple replied:
Yeah, all right. That is where I got this implementation from, but they were slightly different and…yeah, your way makes sense.

jrose-apple replied:
Oh right, the TypeChecker one diagnoses if the pointer types can't be found. I don't really think that's so important, though, do you? A bunch of things break if you have a busted stdlib.

jrose-apple replied:
A little weird to special case AUMP there, though. I'd probably just drop the assert entirely.

brentdax replied:
I think you can justify special-casing AUMP in other places because it genuinely isn’t expected to always exist while the others all should unless the standard library is broken, but dropping the assert would be fine too.

brentdax commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -75,13 +105,16 @@ namespace {
       Boolean,
 
+      /// The source type is 'NSUInteger'.
+      NSUInteger,
```
Any reason why you moved this case?

jrose-apple replied:
All the specific types cases are now together, that's all. (Also, all the nullable cases happen to be together too, but I wasn't specifically optimizing for that.)

brentdax commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -409,9 +443,7 @@ namespace {
-        return {Impl.getNamedSwiftTypeSpecialization(Impl.getStdlibModule(),
-                                                     "UnsafePointer",
-                                                     pointeeType),
+        return {getPointerTo(pointeeType, PTK_UnsafePointer),
```
Oh, these are really nice cleanups.

brentdax commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1156,208 +1195,246 @@ static bool isNSString(Type type) {
-      return Type();
+  // FIXME: Avoid string comparison by caching this identifier.
+  if (elementClass->getName().str() !=
+        impl.SwiftContext.getSwiftName(KnownFoundationEntity::NSError))
```
`getName()` returns an `Identifier` here, doesn't it? So can't you use `getSwiftId()` here instead of `getSwiftName()` to avoid stringifying?

jrose-apple replied:
This code predates `getSwiftId`. Honestly not sure if doing the string compare every time is slower than the time to intern the string anyway, if it's not going to be cached locally per the (ancient) FIXME.

brentdax commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1156,208 +1195,246 @@ static bool isNSString(Type type) {
-    // FIXME: Avoid string comparison by caching this identifier.
-    if (elementClass->getName().str() !=
-          impl.SwiftContext.getSwiftName(KnownFoundationEntity::NSError))
-      return Type();
```
Ah, I see, this code was already here and might have predated `getSwiftId()`.

Come to think of it, I think there might be an `ASTContext::getNSErrorDecl()`. Can you use that to abridge some of this logic?

jrose-apple replied:
For now I think I'd like to leave it as is; the code can always be improved further in the future.

jrose-apple committed:
[AST] Add Type::wrapInPointer(PointerTypeKind)

...replacing a similar utility in the type checker and one I just
added in ClangImporter earlier in this patch series. This is an
ever-so-slight regression in functionality because a missing pointer
declaration used to print a diagnostic and now it will just crash on
an invalid standard library, but I don't think we should be optimizing
for an invalid standard library anyway.

jrose-apple commented:
Updated for Brent's comments, also threw in one more enhancement.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

beccadax reviewed (approved):
LGTM!
